README.md
# FTFoldingTableView

[![Twitter](https://img.shields.io/badge/twitter-@liufengting-blue.svg?style=flat)](http://twitter.com/liufengting) 
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/liufengting/FTFoldingTableView/master/LICENSE)
[![Version](https://img.shields.io/cocoapods/v/FTFoldingTableView.svg?style=flat)](http://cocoapods.org/pods/FTFoldingTableView)
[![Platform](https://img.shields.io/cocoapods/p/FTFoldingTableView.svg?style=flat)](http://cocoapods.org/pods/FTFoldingTableView)
[![CocoaPods](https://img.shields.io/cocoapods/dt/FTFoldingTableView.svg?maxAge=2592000)](http://cocoapods.org/pods/FTFoldingTableView)
[![CI Status](http://img.shields.io/travis/liufengting/FTFoldingTableView.svg?style=flat)](https://travis-ci.org/liufengting/FTFoldingTableView)
[![GitHub stars](https://img.shields.io/github/stars/liufengting/FTFoldingTableView.svg)](https://github.com/liufengting/FTFoldingTableView/stargazers)

FTFoldingTableView. Folding cells with simple delegates, and change UI looks in a few lines of code.


# ScreenShots

| Style	| Left	| Right	|
|:-------------|:-------------:|:-------------:|
| fold | <img src="/Screenshots/Screenshots1.png" width="320"/> | <img src="/Screenshots/Screenshots3.png" width="320"/> |
| show | <img src="/Screenshots/Screenshots2.png" width="320"/> | <img src="/Screenshots/Screenshots4.png" width="320"/> |

# Usage

## Required Methods

* preferred ArrowPosition

```objective-c
- (FTFoldingSectionHeaderArrowPosition)perferedArrowPositionForFTFoldingTableView:(FTFoldingTableView *)ftTableView
{
    return self.arrowPosition;
}
```
* numberOfSectionForFTFoldingTableView

```objective-c
- (NSInteger )numberOfSectionForFTFoldingTableView:(FTFoldingTableView *)ftTableView
{
    return 5;
}

```
* numberOfRowsInSection

```objective-c
- (NSInteger )ftFoldingTableView:(FTFoldingTableView *)ftTableView numberOfRowsInSection:(NSInteger )section
{
    return 3;
}
```
* heightForHeaderInSection

```objective-c
- (CGFloat )ftFoldingTableView:(FTFoldingTableView *)ftTableView heightForHeaderInSection:(NSInteger )section
{
    return 60;
}
```
* heightForRowAtIndexPath

```objective-c
- (CGFloat )ftFoldingTableView:(FTFoldingTableView *)ftTableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 50;
}
```
* titleForHeaderInSection

```objective-c
- (NSString *)ftFoldingTableView:(FTFoldingTableView *)ftTableView titleForHeaderInSection:(NSInteger)section
{
    return [NSString stringWithFormat:@"Section %ld",section];
}
```
* cellForRowAtIndexPath

```objective-c
- (UITableViewCell *)ftFoldingTableView:(FTFoldingTableView *)ftTableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [ft_tableView dequeueReusableCellWithIdentifier:DemoTableViewIdentifier forIndexPath:indexPath];
    
    
    return cell;
}
```
* didSelectRowAtIndexPath

```objective-c
- (void )ftFoldingTableView:(FTFoldingTableView *)ftTableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [ftTableView deselectRowAtIndexPath:indexPath animated:YES];
    
    
}
```
## Optional Methods

* descriptionForHeaderInSection

```objective-c
- (NSString *)ftFoldingTableView:(FTFoldingTableView *)ftTableView descriptionForHeaderInSection:(NSInteger )section
{
    return @"description";
}
```
* arrowImageForSection

```objective-c
- (UIImage *)ftFoldingTableView:(FTFoldingTableView *)ftTableView arrowImageForSection:(NSInteger )section
{
    return [UIImage imageNamed:@"Arrowhead"];
}
```
* backgroundColorForHeaderInSection

```objective-c
- (UIColor *)ftFoldingTableView:(FTFoldingTableView *)ftTableView backgroundColorForHeaderInSection:(NSInteger )section
{
    return [UIColor grayColor];
}
```
* fontForTitleInSection

```objective-c
- (UIFont *)ftFoldingTableView:(FTFoldingTableView *)ftTableView fontForTitleInSection:(NSInteger )section
{
    return [UIFont boldSystemFontOfSize:16];
}
```
* fontForDescriptionInSection

```objective-c
- (UIFont *)ftFoldingTableView:(FTFoldingTableView *)ftTableView fontForDescriptionInSection:(NSInteger )section
{
    return [UIFont systemFontOfSize:13];
}
```
* textColorForTitleInSection

```objective-c
- (UIColor *)ftFoldingTableView:(FTFoldingTableView *)ftTableView textColorForTitleInSection:(NSInteger )section
{
    return [UIColor whiteColor];
}
```
* textColorForDescriptionInSection

```objective-c
- (UIColor *)ftFoldingTableView:(FTFoldingTableView *)ftTableView textColorForDescriptionInSection:(NSInteger )section
{
    return [UIColor whiteColor];
}
```

# Installation

## Manually

* clone this repo.
* Simply drop the '/FTFoldingTableViewLib' folder into your project.
* Enjoy!

## Cocoapods

FTFoldingTableView is available through [CocoaPods](http://cocoapods.org). To install it, simply add the following line to your Podfile:

```ruby
pod 'FTFoldingTableView'
```



# Need to know

* I used two days to finish this project, and then uploaded it to Cocoapods, so I didn't have enough time to test it;if you have any questions, please let me know. I'll fill this README in the near future.




ChangeLog.md
# FTFoldingTableView ChangeLog

* 1.0.0 (2016.06.28) upload to Cocoapods
* 1.0.5 (2016.08.01) supports portrait and landscrape
* 1.0.6 (2017.02.03) add will fold/open delegate
* 1.0.8 (2018.01.11) add didDeselectRowAtIndexPath delegate
* 1.0.9 (2018.01.15) add delegate methods and public methods
* 1.1.0 (2019.03.09) just made some small changes





Demo/FTFoldingTableView/ViewController.m
//
//  ViewController.m
//  FTFoldingTableView
//
//  Created by liufengting on 16/6/20.
//  Copyright © 2016 LiuFengting <https://github.com/liufengting>. All rights reserved.
//

#import "ViewController.h"
#import "FTFoldingTableView.h"

static NSString * const SectionHeaderViewIdentifier    = @"SectionHeaderViewIdentifier";
static NSString * const DemoTableViewIdentifier        = @"DemoTableViewIdentifier";

@interface ViewController () <FTFoldingTableViewDelegate>

@property (weak, nonatomic) IBOutlet FTFoldingTableView *ft_tableView;
@property (nonatomic, assign) FTFoldingSectionHeaderArrowPosition arrowPosition;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    self.automaticallyAdjustsScrollViewInsets = NO;
    
    self.arrowPosition = FTFoldingSectionHeaderArrowPositionLeft;
    
    self.ft_tableView.foldingDelegate = self;
}

- (IBAction)changeStyle:(UIBarButtonItem *)sender
{
    self.arrowPosition = [NSNumber numberWithBool:(![NSNumber numberWithInteger:self.arrowPosition].boolValue)].integerValue;
    [self.ft_tableView reloadData];

    
//  foldAllSections
//    [self.ft_tableView foldAllSections];
    
//  expandAllSections
//    [self.ft_tableView expandAllSections];

//  test one section
//    NSInteger section = 0;
//    FTFoldingSectionState state = [self.ft_tableView foldingStateForSection:section];
//    if (state == FTFoldingSectionStateFold) {
//        [self.ft_tableView expandSection:section];
//    }else{
//        [self.ft_tableView foldSection:section];
//    }
}

#pragma mark - FTFoldingTableViewDelegate / required

- (FTFoldingSectionHeaderArrowPosition)perferedArrowPositionForFTFoldingTableView:(FTFoldingTableView *)ftTableView
{
    return self.arrowPosition;
}
- (NSInteger)numberOfSectionForFTFoldingTableView:(FTFoldingTableView *)ftTableView
{
    return 5;
}
- (NSInteger)ftFoldingTableView:(FTFoldingTableView *)ftTableView numberOfRowsInSection:(NSInteger)section
{
    return 3;
}
- (CGFloat)ftFoldingTableView:(FTFoldingTableView *)ftTableView heightForHeaderInSection:(NSInteger)section
{
    return 60.f;
}
- (CGFloat)ftFoldingTableView:(FTFoldingTableView *)ftTableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 50.f;
}
- (NSString *)ftFoldingTableView:(FTFoldingTableView *)ftTableView titleForHeaderInSection:(NSInteger)section
{
    return [NSString stringWithFormat:@"Section %ld",section];
}
- (UITableViewCell *)ftFoldingTableView:(FTFoldingTableView *)ftTableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell *cell = [self.ft_tableView dequeueReusableCellWithIdentifier:DemoTableViewIdentifier forIndexPath:indexPath];
    
    
    return cell;
}
- (void)ftFoldingTableView:(FTFoldingTableView *)ftTableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [ftTableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - FTFoldingTableViewDelegate / optional delegates, change almost everything using delegate

- (void)ftFoldingTableView:(FTFoldingTableView *)ftTableView willChangeToSectionState:(FTFoldingSectionState)sectionState section:(NSInteger)section
{
    NSLog(@"section: %ld is about to %@", section, sectionState == FTFoldingSectionStateFold ? @"close" : @"open");
}
- (void)ftFoldingTableView:(FTFoldingTableView *)ftTableView didChangeToSectionState:(FTFoldingSectionState)sectionState section:(NSInteger)section
{
    NSLog(@"section: %ld is now %@", section, sectionState == FTFoldingSectionStateFold ? @"closed" : @"opened");
}

- (NSString *)ftFoldingTableView:(FTFoldingTableView *)ftTableView descriptionForHeaderInSection:(NSInteger)section
{
    return @"description";
}

//- (UIImage *)ftFoldingTableView:(FTFoldingTableView *)ftTableView arrowImageForSection:(NSInteger)section
//{
//    return [UIImage imageNamed:@"Arrowhead"];
//}
//- (UIColor *)ftFoldingTableView:(FTFoldingTableView *)ftTableView backgroundColorForHeaderInSection:(NSInteger)section
//{
//    return [UIColor grayColor];
//}
//- (UIFont *)ftFoldingTableView:(FTFoldingTableView *)ftTableView fontForTitleInSection:(NSInteger)section
//{
//    return [UIFont boldSystemFontOfSize:16];
//}
//- (UIFont *)ftFoldingTableView:(FTFoldingTableView *)ftTableView fontForDescriptionInSection:(NSInteger)section
//{
//    return [UIFont systemFontOfSize:13];
//}
//- (UIColor *)ftFoldingTableView:(FTFoldingTableView *)ftTableView textColorForTitleInSection:(NSInteger)section
//{
//    return [UIColor whiteColor];
//}
//- (UIColor *)ftFoldingTableView:(FTFoldingTableView *)ftTableView textColorForDescriptionInSection:(NSInteger)section
//{
//    return [UIColor whiteColor];
//}

@end
